src/os/bluestore/BlueStore.h

```diff
@@ -1111,11 +1118,18 @@ class BlueStore : public ObjectStore,
 
     void flush();
     void get() {
-      ++nref;
+      if (++nref == 2 && s != nullptr && !pinned && !pinning) {
```

Comment:
I thought I left this comment last week, but apparently it didn't get saved:

The use of atomics here isn't safe - consider the execution of 3 threads, a b and c, with nref=2 and pinned=true initially for a particular onode o:
```
a: put() -> n = --nref = 1
b: get() -> ++nref = 2
c: put() -> n = --nref = 1
a: pinned == true && !pinning -> unpin()
c: pinned == true && !pinning -> unpin()
a: unpin() -> pin_list.remove(pin_list.iterator_to(o))
c: unpin() -> invalid arg for pin_list.iterator_to - o is not in pin_list
```
The opposite could happen with the onode getting added to pin_list more than once as well.

An easy fix would be making get() and put() take s->lock() every time, making nref an int instead of atomic, and removing the pinning/pinned atomics. In the other places pinned is used, _rm() and _touch(), the cache lock is already held by the callers. This has the disadvantage that you're now taking a lock for every onode get()/put().

I think you could avoid this by changing pinned to a bool, and protecting it with s->lock. This way you could only check it when you might need to pin/unpin, but you can't alter the pin_list without changing pinned. This could be:

```
void get() {
  if (++nref == 2 && s != nullptr) {
    std::lock_guard l(s->lock);
    if (!pinned)
      s->pin();
  }
}
void put() {
  int n = --nref;
  if (n == 1 && s != nullptr) {
    std::lock_guard l(s->lock);
    if (pinned)
      s->unpin();
  }
}
```

Another way to approach this would be trying to change all the callers of get()/put() for onodes, but you'd still need some locking to protect the pin_list, so I think that would be worse by spreading the logic to the callers.

Reply:
> I thought I left this comment last week, but apparently it didn't get saved:
> 
> The use of atomics here isn't safe - consider the execution of 3 threads, a b and c, with nref=2 and pinned=true initially for a particular onode o:
> 
> ```
> a: put() -> n = --nref = 1
> b: get() -> ++nref = 2
> c: put() -> n = --nref = 1
> a: pinned == true && !pinning -> unpin()
> c: pinned == true && !pinning -> unpin()
> a: unpin() -> pin_list.remove(pin_list.iterator_to(o))
> c: unpin() -> invalid arg for pin_list.iterator_to - o is not in pin_list
> ```
> 
> The opposite could happen with the onode getting added to pin_list more than once as well.
> 
> An easy fix would be making get() and put() take s->lock() every time, making nref an int instead of atomic, and removing the pinning/pinned atomics. In the other places pinned is used, _rm() and _touch(), the cache lock is already held by the callers. This has the disadvantage that you're now taking a lock for every onode get()/put().
> 
> I think you could avoid this by changing pinned to a bool, and protecting it with s->lock. This way you could only check it when you might need to pin/unpin, but you can't alter the pin_list without changing pinned. This could be:
> 
> ```
> void get() {
>   if (++nref == 2 && s != nullptr) {
>     std::lock_guard l(s->lock);
>     if (!pinned)
>       s->pin();
>   }
> }
> void put() {
>   int n = --nref;
>   if (n == 1 && s != nullptr) {
>     std::lock_guard l(s->lock);
>     if (pinned)
>       s->unpin();
>   }
> }
> ```
> 
> Another way to approach this would be trying to change all the callers of get()/put() for onodes, but you'd still need some locking to protect the pin_list, so I think that would be worse by spreading the logic to the callers.

Excellent catch Josh!  That was totally braindead and I disavow whatever I was thinking last week. :)  As we discussed on IRC I've gotten rid of pinning entirely, changed pinned to a bool, and moved the associated conditional check into _pin()/_unpin() so we can keep the same locking wrapper convention that we use in other places in the cache.  A quick performance test doesn't show any slowdown:

#30964 default old 4K randwrite: 39732.5
#39064 default new 4K randwrite: 39823.3